### PR TITLE
Unify stret detection between the generator and platform assemblies, and fix stret detection on watchOS. Fixes #44709.

### PR DESCRIPTION
**Pull request by rolfbjarne**
Unify the code to determine whether a particular return type requires a
stret signature or not between the generator and platform assemblies.

Also fix the stret detection for armv7k, whose calling convention is not
identical to armv7(s): there's the concept of homogeneous structures, which
contains multiple elements of only one type, and which is sometimes passed
in registers on armv7k.

The tests I wrote for this bug found several other issues, which are also fixed.

https://bugzilla.xamarin.com/show_bug.cgi?id=44709

### CODE REVIEW

File: src/MetalPerformanceShaders/MPSDefs.cs [hunk 1]
```diff
@@ -81,10 +83,19 @@ public struct MPSScaleTransform {
...
 	// MPSImageHistogram.h
+	[StructLayout (LayoutKind.Explicit)]
 	public struct MPSImageHistogramInfo {
+		[FieldOffset (0)]
 		public nuint NumberOfHistogramEntries;
+#if XAMCORE_2_0
+		[FieldOffset (8)]
+#else
+		[FieldOffset (4)]
+#endif
```
Comment (spouliot) [inline src/MetalPerformanceShaders/MPSDefs.cs]: What's the story here ? :-)
Only XM classic won't defined `XAMCORE_2_0` and IIRC it does not support Metal, so it looks like _dead_ code.

Comment (rolfbjarne) [inline src/MetalPerformanceShaders/MPSDefs.cs]: Oops, that's wrong, it's supposed be a 32-bit/64-bit ifdef, not Classic/Unified.

Comment (rolfbjarne) [inline src/MetalPerformanceShaders/MPSDefs.cs]: Fixed.

**Comment by rolfbjarne** [comment]
build

**Comment by rolfbjarne** [comment]
3 test runs failed, none with the same test. Looks like 3 random test failures 😒

**Pull request merged**